File: mm-kt/src/main/kotlin/io/github/metamessage/ir/ValueType.kt
package io.github.metamessage.ir

enum class ValueType {
    UNKNOWN,
    DOC,
    VEC,
    ARR,
    OBJ,
    MAP,
    STR,
    BYTES,
    BOOL,
    I,
    I8,
    I16,
    I32,
    I64,
    U,
    U8,
    U16,
    U32,
    U64,
    F32,
    F64,
    BIGINT,
    DATETIME,
    DATE,
    TIME,
    UUID,
    DECIMAL,
    IP,
    URL,
    EMAIL,
    ENUM,
    IMAGE,
    VIDEO;

    fun code(): Byte = ordinal.toByte()

    override fun toString(): String =
            when (this) {
                UNKNOWN -> "unknown"
                DOC -> "doc"
                VEC -> "vec"
                ARR -> "arr"
                OBJ -> "obj"
                MAP -> "map"
                STR -> "str"
                BYTES -> "bytes"
                BOOL -> "bool"
                I -> "i"
                I8 -> "i8"
                I16 -> "i16"
                I32 -> "i32"
                I64 -> "i64"
                U -> "u"
                U8 -> "u8"
                U16 -> "u16"
                U32 -> "u32"
                U64 -> "u64"
                F32 -> "f32"
                F64 -> "f64"
                BIGINT -> "bigint"
                DATETIME -> "datetime"
                DATE -> "date"
                TIME -> "time"
                UUID -> "uuid"
                DECIMAL -> "decimal"
                IP -> "ip"
                URL -> "url"
                EMAIL -> "email"
                ENUM -> "enum"
                IMAGE -> "image"
                VIDEO -> "video"
            }

    companion object {
        fun fromCode(b: Int): ValueType {
            val values = entries
            return if (b < 0 || b >= values.size) UNKNOWN else values[b]
        }

        private val WIRE: Map<String, ValueType> = run {
            val map = mutableMapOf<String, ValueType>()
            for (v in entries) {
                map[v.toString()] = v
            }
            map
        }

        fun parseWireName(s: String): ValueType {
            if (s.isEmpty()) return UNKNOWN
            return WIRE[s.trim().lowercase()] ?: UNKNOWN
        }
    }
}

File: mm-php/src/io/metamessage/ir/ValueType.php
<?php

namespace io\metamessage\ir;

enum ValueType
{
    case UNKNOWN;
    case DOC;
    case VEC;
    case ARR;
    case OBJ;
    case MAP;
    case STR;
    case BYTES;
    case BOOL;
    case I;
    case I8;
    case I16;
    case I32;
    case I64;
    case U;
    case U8;
    case U16;
    case U32;
    case U64;
    case F32;
    case F64;
    case BIGINT;
    case DATETIME;
    case DATE;
    case TIME;
    case UUID;
    case DECIMAL;
    case IP;
    case URL;
    case EMAIL;
    case ENUM;
    case IMAGE;
    case VIDEO;

    public function code(): int
    {
        return array_search($this, self::cases());
    }

    public static function fromCode(int $code): self
    {
        $cases = self::cases();
        return $cases[$code] ?? self::UNKNOWN;
    }

    public function wireName(): string
    {
        return match ($this) {
            self::UNKNOWN => 'unknown',
            self::DOC => 'doc',
            self::VEC => 'vec',
            self::ARR => 'arr',
            self::OBJ => 'obj',
            self::MAP => 'map',
            self::STR => 'str',
            self::BYTES => 'bytes',
            self::BOOL => 'bool',
            self::I => 'i',
            self::I8 => 'i8',
            self::I16 => 'i16',
            self::I32 => 'i32',
            self::I64 => 'i64',
            self::U => 'u',
            self::U8 => 'u8',
            self::U16 => 'u16',
            self::U32 => 'u32',
            self::U64 => 'u64',
            self::F32 => 'f32',
            self::F64 => 'f64',
            self::BIGINT => 'bigint',
            self::DATETIME => 'datetime',
            self::DATE => 'date',
            self::TIME => 'time',
            self::UUID => 'uuid',
            self::DECIMAL => 'decimal',
            self::IP => 'ip',
            self::URL => 'url',
            self::EMAIL => 'email',
            self::ENUM => 'enum',
            self::IMAGE => 'image',
            self::VIDEO => 'video',
        };
    }

    public static function parseWireName(string $s): self
    {
        if (empty($s)) {
            return self::UNKNOWN;
        }
        $key = strtolower(trim($s));
        if ($key === 'struct') {
            return self::OBJ;
        }
        $wireNameMap = array_reduce(self::cases(), function ($map, $case) {
            $map[strtolower($case->wireName())] = $case;
            return $map;
        }, []);
        return $wireNameMap[$key] ?? self::UNKNOWN;
    }
}
